README.md
# Sibila

Sibila es una librería que permite realizar predicciones (forecasts) de manera sencilla, utilizando por detrás el servicio de Forecasting de AWS. 

### Instalación
```
python3 -m pip install boto3 && \
    python3 -m pip install --index-url http://pypi.ml.com/simple/ \
    --trusted-host pypi.ml.com sibila
```

### Requisitos
- Credenciales de AWS
- Dataset de entrenamiento

**Obtención de las credenciales:**

 1. Generar una nueva solicitud de acceso a AWS:
    
    1.  [Ingresar a SHIELD](https://shield.adminml.com/)
    
    2.  Crear una solicitud de tipo "AWS - Acceso vía Bastión"
    
		- Cuenta: 628956477585

		- Rol: CrossAccountManager-BI-Forecasting

		- Tipo de acceso: Permanente
		⚠️ **Atención:** Si no está en la lista de roles, seleccionar “NO ENCONTRE EL ROL EN LA LISTA” y escribir el nombre del rol el campo de texto del form de Shield)

  2. Seguir las [instrucciones](https://github.com/mercadolibre/fury_aws-bastion-cli-client/blob/master/README.md) de uso del cliente de **bastion** para generar credenciales temporarias

### Uso

Se puede encontrar un ejemplo de uso básico en esta [notebook](./docs/basic_example.ipynb), y más información específica sobre las entidades y métodos que ofrece la librería [aquí](./docs).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.2.1] - 25 May 2021

### Fixed

- `Project.get_dataset()` is now able to return a dataset by using the most recent import job

## [1.2.0] - 6 Apr 2021

### Added

- `forecast_dimensions` as a `Predictor` parameter

## [1.1.0] - 9 Feb 2021

### Added

- `Predictor.info` property that updates predictor information and returns an info dict.

- `Predictor.wait_for_training()` so that users are able to time the training duration.

## [1.0.0] - 9 Feb 2021

### Changed

- `Project` now requires LDAP user for authentication.

## [0.1.1] - 4 Feb 2021

### Changed

- The predictor now waits for all datasets to be `ACTIVE` before creating a forecast.


## [0.1.0] - 28 Jan 2021

### Added

- `supplementary_features` as an optional parameter in `train_new_predictor()`


<!-- Safe to delete, just as example -->
## [0.0.1] - example - 04 Nov 2020

### Added

- New feature added

### Changed

- Fixed bug
- Enhanced feature
- Refactored code

### Removed

- Unnecessary file


tests/__init__.py
version = "0.0.1-dev0"


tests/conftest.py
import pytest
import datetime
from unittest.mock import patch
import requests_mock
from http import HTTPStatus

from sibila.project import Project


def tzlocal():
    return


@pytest.fixture()
def project(
    describe_dataset_group_response,
    list_predictors_response,
    describe_dataset_import_job_response,
    describe_dataset_response,
    describe_predictor_response,
):
    with patch("sibila.project.AWSHandler") as handler:
        mocked_handler = handler.return_value
        mocked_handler.forecast.describe_dataset_group.return_value = (
            describe_dataset_group_response
        )
        mocked_handler.forecast.describe_dataset.return_value = (
            describe_dataset_response
        )
        mocked_handler.forecast.list_predictors.return_value = list_predictors_response
        mocked_handler.forecast.describe_dataset_import_job.return_value = (
            describe_dataset_import_job_response
        )
        mocked_handler.forecast.describe_predictor.return_value = (
            describe_predictor_response
        )

        # For all the responses below anything with a STATUS passes all tests
        mocked_handler.forecast.describe_forecast.return_value = (
            describe_predictor_response
        )
        mocked_handler.forecast.describe_forecast_export_job.return_value = (
            describe_predictor_response
        )
        mocked_handler.forecast.describe_forecast_backtest_export_job.return_value = (
            describe_predictor_response
        )
        mocked_handler.forecast.describe_predictor_backtest_export_job.return_value = (
            describe_predictor_response
        )

        mocked_handler.team = "EQUIPO_DE_PRUEBA_1"
        with requests_mock.Mocker() as request_mock:
            request_mock.post(
                "https://internal-api.mercadolibre.com/data-team/auth/login",
                status_code=HTTPStatus.NO_CONTENT,
            )
            project = Project(
                name="nombre-de_prueba-1",
                team="equipo_de-prueba-1",
                username="username",
                password="password",
                aws_access_key_id="a",
                aws_secret_access_key="b",
                aws_session_token="c",
                region_name="region",
            )
            return project


@pytest.fixture()
def list_predictors_response():
    return {
        "Predictors": [
            {
                "PredictorArn": "arn:aws:forecast:us-east-1:628956477585:predictor/BI_ML_CROSS__ARIMA_demanda",
                "PredictorName": "BI_ML_CROSS__ARIMA_demanda",
                "DatasetGroupArn": "arn:aws:forecast:us-east-1:628956477585:dataset-group/BI_ML_CROSS__demanda_point",
                "Status": "ACTIVE",
                "CreationTime": datetime.datetime(
                    2020, 12, 22, 17, 47, 41, 4000, tzinfo=tzlocal()
                ),
                "LastModificationTime": datetime.datetime(
                    2020, 12, 22, 19, 40, 32, 426000, tzinfo=tzlocal()
                ),
            },
            {
                "PredictorArn": "arn:aws:forecast:us-east-1:628956477585:predictor/forecast_ventas_all",
                "PredictorName": "forecast_ventas_all",
                "DatasetGroupArn": "arn:aws:forecast:us-east-1:628956477585:dataset-group/MGG_MLA_CE_ALL",
                "Status": "CREATE_FAILED",
                "Message": "Items found in the TARGET_TIME_SERIES dataset are missing from the ITEM_METADATA dataset. A total of 153 items are missing from the ITEM_METADATA dataset.",
                "CreationTime": datetime.datetime(
                    2020, 11, 27, 9, 44, 23, 249000, tzinfo=tzlocal()
                ),
                "LastModificationTime": datetime.datetime(
                    2020, 11, 27, 9, 54, 12, 814000, tzinfo=tzlocal()
                ),
            },
            {
                "PredictorArn": "arn:aws:forecast:us-east-1:628956477585:predictor/ventas_publicaiones",
                "PredictorName": "ventas_publicaiones",
                "DatasetGroupArn": "arn:aws:forecast:us-east-1:628956477585:dataset-group/Ventas_metadata",
                "Status": "ACTIVE",
                "CreationTime": datetime.datetime(
                    2020, 11, 23, 22, 19, 18, 796000, tzinfo=tzlocal()
                ),
                "LastModificationTime": datetime.datetime(
                    2020, 11, 24, 0, 5, 36, 620000, tzinfo=tzlocal()
                ),
            },
            {
                "PredictorArn": "arn:aws:forecast:us-east-1:628956477585:predictor/solo_ventas_est",
                "PredictorName": "solo_ventas_est",
                "DatasetGroupArn": "arn:aws:forecast:us-east-1:628956477585:dataset-group/MGG_MLA_CE_ONLY_SALES",
                "Status": "ACTIVE",
                "CreationTime": datetime.datetime(
                    2020, 11, 23, 18, 30, 38, 380000, tzinfo=tzlocal()
                ),
                "LastModificationTime": datetime.datetime(
                    2020, 11, 23, 20, 6, 8, 993000, tzinfo=tzlocal()
                ),
            },
            {
                "PredictorArn": "arn:aws:forecast:us-east-1:628956477585:predictor/ventas_est",
                "PredictorName": "ventas_est",
                "DatasetGroupArn": "arn:aws:forecast:us-east-1:628956477585:dataset-group/MGG_MLA_CE",
                "Status": "ACTIVE",
                "CreationTime": datetime.datetime(
                    2020, 11, 23, 16, 13, 49, 621000, tzinfo=tzlocal()
                ),
                "LastModificationTime": datetime.datetime(
                    2020, 11, 23, 16, 44, 35, 900000, tzinfo=tzlocal()
                ),
            },
            {
                "PredictorArn": "arn:aws:forecast:us-east-1:628956477585:predictor/MLM_predictor_till_august",
                "PredictorName": "MLM_predictor_till_august",
                "DatasetGroupArn": "arn:aws:forecast:us-east-1:628956477585:dataset-group/item_variation_forecast",
                "Status": "ACTIVE",
                "CreationTime": datetime.datetime(
                    2020, 11, 19, 13, 35, 50, 69000, tzinfo=tzlocal()
                ),
                "LastModificationTime": datetime.datetime(
                    2020, 11, 19, 16, 47, 6, 477000, tzinfo=tzlocal()
                ),
            },
            {
                "PredictorArn": "arn:aws:forecast:us-east-1:628956477585:predictor/my_modelo",
                "PredictorName": "my_modelo",
                "DatasetGroupArn": "arn:aws:forecast:us-east-1:628956477585:dataset-group/mlseeds_forecast",
                "Status": "ACTIVE",
                "CreationTime": datetime.datetime(
                    2020, 10, 15, 15, 28, 32, 159000, tzinfo=tzlocal()
                ),
                "LastModificationTime": datetime.datetime(
                    2020, 10, 15, 15, 45, 17, 464000, tzinfo=tzlocal()
                ),
            },
        ],
        "ResponseMetadata": {
            "RequestId": "32506860-76f1-4d64-ad39-527e3f3fc2bd",
            "HTTPStatusCode": 200,
            "HTTPHeaders": {
                "content-type": "application/x-amz-json-1.1",
                "date": "Mon, 28 Dec 2020 13:24:35 GMT",
                "x-amzn-requestid": "32506860-76f1-4d64-ad39-527e3f3fc2bd",
                "content-length": "2335",
                "connection": "keep-alive",
            },
            "RetryAttempts": 0,
        },
    }


@pytest.fixture()
def describe_dataset_group_response():
    return {
        "DatasetGroupName": "BI_ML_CROSS__demanda_point",
        "DatasetGroupArn": "arn:aws:forecast:us-east-1:628956477585:dataset-group/BI_ML_CROSS__demanda_point",
        "DatasetArns": [
            "arn:aws:forecast:us-east-1:628956477585:dataset/BI_ML_CROSS__demanda_point__TTS"
        ],
        "Domain": "CUSTOM",
        "Status": "ACTIVE",
        "CreationTime": datetime.datetime(
            2020, 12, 22, 15, 45, 44, 878000, tzinfo=tzlocal()
        ),
        "LastModificationTime": datetime.datetime(
            2020, 12, 22, 17, 47, 58, 522000, tzinfo=tzlocal()
        ),
        "ResponseMetadata": {
            "RequestId": "2cf84b7c-3c05-4227-94d1-4195ba7c4d71",
            "HTTPStatusCode": 200,
            "HTTPHeaders": {
                "content-type": "application/x-amz-json-1.1",
                "date": "Mon, 28 Dec 2020 15:09:14 GMT",
                "x-amzn-requestid": "2cf84b7c-3c05-4227-94d1-4195ba7c4d71",
                "content-length": "356",
                "connection": "keep-alive",
            },
            "RetryAttempts": 0,
        },
    }


@pytest.fixture()
def describe_dataset_import_job_response():
    return {
        "DatasetImportJobName": "demanada_2018_a_2020_import",
        "DatasetImportJobArn": "arn:aws:forecast:us-east-1:628956477585:dataset-import-job/demanada_2018_a_2020/demanada_2018_a_2020_import",
        "DatasetArn": "arn:aws:forecast:us-east-1:628956477585:dataset/demanada_2018_a_2020",
        "TimestampFormat": "yyyy-MM-dd hh:mm:ss",
        "UseGeolocationForTimeZone": False,
        "DataSource": {
            "S3Config": {
                "Path": "s3://bi-ml-forecasting-data/input/mi_equipo/forecast_demanda_producto/TARGET_TIME_SERIES/demanada_2018_a_2020.csv",
                "RoleArn": "arn:aws:iam::628956477585:role/BI-Forecast",
            }
        },
        "FieldStatistics": {
            "item_id": {"Count": 23973, "CountDistinct": 3, "CountNull": 0},
            "target_value": {
                "Count": 23973,
                "CountDistinct": 4818,
                "CountNull": 0,
                "CountNan": 0,
                "Min": "0.0",
                "Max": "212.27197346600326",
                "Avg": 50.44732317068067,
                "Stddev": 38.72169238224657,
            },
            "timestamp": {
                "Count": 23973,
                "CountDistinct": 7991,
                "CountNull": 0,
                "Min": "2014-01-01T01:00:00Z",
                "Max": "2014-11-29T23:00:00Z",
            },
        },
        "DataSize": 0.0010688817128539085,
        "Status": "ACTIVE",
        "CreationTime": datetime.datetime(
            2020, 12, 11, 16, 2, 26, 674000, tzinfo=tzlocal()
        ),
        "LastModificationTime": datetime.datetime(
            2020, 12, 11, 16, 37, 4, 803000, tzinfo=tzlocal()
        ),
        "ResponseMetadata": {
            "RequestId": "2d6c10c9-04e6-43b4-9426-331df24be55d",
            "HTTPStatusCode": 200,
            "HTTPHeaders": {
                "content-type": "application/x-amz-json-1.1",
                "date": "Mon, 14 Dec 2020 13:15:20 GMT",
                "x-amzn-requestid": "2d6c10c9-04e6-43b4-9426-331df24be55d",
                "content-length": "1238",
                "connection": "keep-alive",
            },
            "RetryAttempts": 0,
        },
    }


@pytest.fixture()
def describe_dataset_response():
    return {
        "DatasetArn": "arn:aws:forecast:us-east-1:628956477585:dataset/generate_periodic_func__70__5_4_0_5__0_4_0_5",
        "DatasetName": "generate_periodic_func__70__5_4_0_5__0_4_0_5",
        "Domain": "RETAIL",
        "DatasetType": "TARGET_TIME_SERIES",
        "DataFrequency": "D",
        "Schema": {
            "Attributes": [
                {"AttributeName": "item_id", "AttributeType": "string"},
                {"AttributeName": "timestamp", "AttributeType": "timestamp"},
                {"AttributeName": "demand", "AttributeType": "float"},
            ]
        },
        "EncryptionConfig": {},
        "Status": "ACTIVE",
        "CreationTime": datetime.datetime(
            2020, 10, 19, 15, 50, 53, 341000, tzinfo=tzlocal()
        ),
        "LastModificationTime": datetime.datetime(
            2020, 10, 19, 17, 28, 41, 166000, tzinfo=tzlocal()
        ),
        "ResponseMetadata": {
            "RequestId": "4c4417b2-a7d9-4fc5-a41a-323ecdc030db",
            "HTTPStatusCode": 200,
            "HTTPHeaders": {
                "content-type": "application/x-amz-json-1.1",
                "date": "Mon, 14 Dec 2020 13:19:39 GMT",
                "x-amzn-requestid": "4c4417b2-a7d9-4fc5-a41a-323ecdc030db",
                "content-length": "543",
                "connection": "keep-alive",
            },
            "RetryAttempts": 0,
        },
    }


@pytest.fixture()
def describe_predictor_response():
    return {
        "PredictorArn": "arn:aws:forecast:us-east-1:628956477585:predictor/predictor_1",
        "PredictorName": "predictor_1",
        "ForecastHorizon": 80,
        "ForecastTypes": ["0.1", "0.5", "0.9"],
        "PerformAutoML": True,
        "TrainingParameters": {
            "context_length": "483",
            "epochs": "100",
            "use_item_metadata": "NONE",
            "use_related_data": "HISTORICAL",
        },
        "EvaluationParameters": {
            "NumberOfBacktestWindows": 1,
            "BackTestWindowOffset": 80,
        },
        "InputDataConfig": {
            "DatasetGroupArn": "arn:aws:forecast:us-east-1:628956477585:dataset-group/testing_func_fit"
        },
        "FeaturizationConfig": {
            "ForecastFrequency": "D",
            "ForecastDimensions": ["D", "B"],
            "Featurizations": [
                {
                    "AttributeName": "demand",
                    "FeaturizationPipeline": [
                        {
                            "FeaturizationMethodName": "filling",
                            "FeaturizationMethodParameters": {
                                "aggregation": "sum",
                                "backfill": "zero",
                                "frontfill": "none",
                                "middlefill": "zero",
                            },
                        }
                    ],
                }
            ],
        },
        "PredictorExecutionDetails": {
            "PredictorExecutions": [
                {
                    "AlgorithmArn": "arn:aws:forecast:::algorithm/ETS",
                    "TestWindows": [
                        {
                            "TestWindowStart": datetime.datetime(
                                2021, 1, 24, 21, 0, tzinfo=tzlocal()
                            ),
                            "TestWindowEnd": datetime.datetime(
                                2021, 4, 14, 21, 0, tzinfo=tzlocal()
                            ),
                            "Status": "ACTIVE",
                        }
                    ],
                },
                {
                    "AlgorithmArn": "arn:aws:forecast:::algorithm/CNN-QR",
                    "TestWindows": [
                        {
                            "TestWindowStart": datetime.datetime(
                                2021, 1, 24, 21, 0, tzinfo=tzlocal()
                            ),
                            "TestWindowEnd": datetime.datetime(
                                2021, 4, 14, 21, 0, tzinfo=tzlocal()
                            ),
                            "Status": "ACTIVE",
                        }
                    ],
                },
                {
                    "AlgorithmArn": "arn:aws:forecast:::algorithm/NPTS",
                    "TestWindows": [
                        {
                            "TestWindowStart": datetime.datetime(
                                2021, 1, 24, 21, 0, tzinfo=tzlocal()
                            ),
                            "TestWindowEnd": datetime.datetime(
                                2021, 4, 14, 21, 0, tzinfo=tzlocal()
                            ),
                            "Status": "ACTIVE",
                        }
                    ],
                },
                {
                    "AlgorithmArn": "arn:aws:forecast:::algorithm/Prophet",
                    "TestWindows": [
                        {
                            "TestWindowStart": datetime.datetime(
                                2021, 1, 24, 21, 0, tzinfo=tzlocal()
                            ),
                            "TestWindowEnd": datetime.datetime(
                                2021, 4, 14, 21, 0, tzinfo=tzlocal()
                            ),
                            "Status": "ACTIVE",
                        }
                    ],
                },
                {
                    "AlgorithmArn": "arn:aws:forecast:::algorithm/ARIMA",
                    "TestWindows": [
                        {
                            "TestWindowStart": datetime.datetime(
                                2021, 1, 24, 21, 0, tzinfo=tzlocal()
                            ),
                            "TestWindowEnd": datetime.datetime(
                                2021, 4, 14, 21, 0, tzinfo=tzlocal()
                            ),
                            "Status": "ACTIVE",
                        }
                    ],
                },
                {
                    "AlgorithmArn": "arn:aws:forecast:::algorithm/Deep_AR_Plus",
                    "TestWindows": [
                        {
                            "TestWindowStart": datetime.datetime(
                                2021, 1, 24, 21, 0, tzinfo=tzlocal()
                            ),
                            "TestWindowEnd": datetime.datetime(
                                2021, 4, 14, 21, 0, tzinfo=tzlocal()
                            ),
                            "Status": "ACTIVE",
                        }
                    ],
                },
            ]
        },
        "DatasetImportJobArns": [
            "arn:aws:forecast:us-east-1:628956477585:dataset-import-job/generate_periodic_func__70__5_4_0_5__0_4_0_5/prueba_2"
        ],
        "AutoMLAlgorithmArns": ["arn:aws:forecast:::algorithm/CNN-QR"],
        "Status": "ACTIVE",
        "CreationTime": datetime.datetime(
            2020, 10, 19, 17, 37, 6, 44000, tzinfo=tzlocal()
        ),
        "LastModificationTime": datetime.datetime(
            2020, 10, 20, 12, 2, 17, 788000, tzinfo=tzlocal()
        ),
        "ResponseMetadata": {
            "RequestId": "05f6d2e6-2b9d-4d7b-aea7-8e1e841181d7",
            "HTTPStatusCode": 200,
            "HTTPHeaders": {
                "content-type": "application/x-amz-json-1.1",
                "date": "Mon, 14 Dec 2020 13:41:28 GMT",
                "x-amzn-requestid": "05f6d2e6-2b9d-4d7b-aea7-8e1e841181d7",
                "content-length": "2008",
                "connection": "keep-alive",
            },
            "RetryAttempts": 0,
        },
    }


sibila/predictor.py
import os
from datetime import datetime

from melitk import logging

from sibila.errors import PredictorError, ResourceError
from sibila.utils import wait_for_resource, download_and_merge, logger


class Predictor:

    AUTO = "AUTO"
    ARIMA = "ARIMA"
    DEEP_AR_PLUS = "Deep_AR_Plus"
    ETS = "ETS"
    NPTS = "NPTS"
    PROPHET = "Prophet"
    CNN_QR = "CNN-QR"

    VALID_ALGORITHMS = (AUTO, ARIMA, DEEP_AR_PLUS, ETS, NPTS, PROPHET, CNN_QR)

    def __init__(
        self,
        aws_handler: "AWSHandler",
        dsg_arn: str,
        name: str,
        algorithm: str,
        horizon: int,
        frequency: str,
        forecast_dimensions: list = None,
        featurizations: list = None,
        supplementary_features: list = None,
        backtest_windows: int = 1,
        window_offset: int = 0,
        hpo: int = False,
        hpo_config: dict = None,
        training_parameters: dict = None,
    ):
        if algorithm not in Predictor.VALID_ALGORITHMS:
            raise PredictorError(f"Valid algorithms are {Predictor.VALID_ALGORITHMS}")

        self.aws_handler = aws_handler
        self.dsg_arn = dsg_arn
        self.name = name
        self.algorithm = algorithm
        self.horizon = horizon
        self.backtest_windows = backtest_windows
        self.frequency = frequency
        self.forecast_dimensions = forecast_dimensions
        self.featurizations = featurizations
        self.supplementary_features = supplementary_features
        self.window_offset = window_offset or horizon
        self.hpo = hpo
        self.hpo_config = hpo_config
        self.training_parameters = training_parameters
        self.perform_automl = algorithm == Predictor.AUTO
        # If HPOConfig is given, HPO must be set to True
        if self.hpo_config:
            self.hpo = True

        # These are set manually either when the predictor is instantiated or upon retrieval.
        self._predictor_name = None
        self.arn = None

    def _create(self):
        """Instantiates the predictor on the AWS forecast service"""

        self._request = {
            "PredictorName": self._predictor_name,
            "ForecastHorizon": self.horizon,
            "PerformAutoML": self.perform_automl,
            "PerformHPO": self.hpo,
            "EvaluationParameters": {
                "NumberOfBacktestWindows": self.backtest_windows,
                "BackTestWindowOffset": self.window_offset,
            },
            "InputDataConfig": {"DatasetGroupArn": self.dsg_arn},
            "FeaturizationConfig": {
                "ForecastFrequency": self.frequency,
            },
            "Tags": [{"Key": "Name", "Value": self.aws_handler.team}],
        }

        if not self.perform_automl:
            self._request["AlgorithmArn"] = (
                "arn:aws:forecast:::algorithm/" + self.algorithm
            )

        if self.featurizations != None:
            self._request["FeaturizationConfig"][
                "Featurizations"
            ] = self.featurizations or [
                {
                    "AttributeName": "target_value",
                    "FeaturizationPipeline": [
                        {
                            "FeaturizationMethodName": "filling",
                            "FeaturizationMethodParameters": {
                                "frontfill": "none",
                                "middlefill": "zero",
                                "backfill": "zero",
                            },
                        }
                    ],
                }
            ]

        if self.supplementary_features != None:
            self._request["InputDataConfig"][
                "SupplementaryFeatures"
            ] = self.supplementary_features

        if self.forecast_dimensions != None:
            self._request["FeaturizationConfig"][
                "ForecastDimensions"
            ] = self.forecast_dimensions

        if self.hpo_config != None:
            self._request["HPOConfig"] = self.hpo_config

        if self.training_parameters != None:
            self._request["TrainingParameters"] = self.training_parameters

        logger.debug(f"Predictor creation request: {self._request}")
        return self.aws_handler.forecast.create_predictor(**self._request)[
            "PredictorArn"
        ]

    def _update(self):
        predictor_info = self.aws_handler.forecast.describe_predictor(
            PredictorArn=self.arn
        )
        try:
            algorithm = predictor_info["AlgorithmArn"].split("/")[-1]
        except KeyError:
            algorithm = predictor_info["AutoMLAlgorithmArns"][0].split("/")[-1]

        self.algorithm = algorithm
        self.horizon = predictor_info["ForecastHorizon"]
        self.frequency = predictor_info["FeaturizationConfig"]["ForecastFrequency"]
        self.forecast_dimensions = predictor_info["FeaturizationConfig"].get(
            "ForecastDimensions"
        )
        self.featurizations = predictor_info["FeaturizationConfig"].get(
            "Featurizations"
        )
        self.supplementary_features = predictor_info["InputDataConfig"].get(
            "SupplementaryFeatures"
        )
        self.backtest_windows = predictor_info["EvaluationParameters"][
            "NumberOfBacktestWindows"
        ]
        self.window_offset = predictor_info["EvaluationParameters"][
            "BackTestWindowOffset"
        ]
        self.hpo = predictor_info.get("PerformHPO")
        self.hpo_config = predictor_info.get("HPOConfig")
        self.training_parameters = predictor_info.get("TrainingParameters")

    @property
    def info(self):
        self._update()
        return {
            "name": self.name,
            "algorithm": self.algorithm,
            "horizon": self.horizon,
            "frequency": self.frequency,
            "backtest_windows": self.backtest_windows,
            "window_offset": self.window_offset,
            "forecast_dimensions": self.forecast_dimensions,
            "featurizations": self.featurizations,
            "supplementary_features": self.supplementary_features,
            "hpo": self.hpo,
            "hpo_config": self.hpo_config,
            "training_parameters": self.training_parameters,
        }

    def wait_for_training(self):
        try:
            wait_for_resource("predictor", self.arn, self.aws_handler)
        except ResourceError as e:  # pragma: no cover
            logger.error(f"There was an error {e} when creating the predictor")
            return

    def metrics(self, destination_dir):

        if not os.path.isdir(destination_dir):  # pragma: no cover
            logger.info(f"Creating directory: {destination_dir}")
            os.mkdir(destination_dir)

        name = "backtest_" + str(datetime.now().timestamp()).replace(".", "_")
        s3_uri = f"{self.aws_handler.s3_uri}/backtests/{name}"

        # Wait for the predictor to be active before trying to retrieve the metrics.
        logger.info("Waiting for the predictor to be ready...")
        try:
            wait_for_resource("predictor", self.arn, self.aws_handler)
        except ResourceError as e:  # pragma: no cover
            logger.error(
                f"The metrics could not be retrieved because there was an error {e} when creating the predictor"
            )
            return

        # Create the backtest export job.
        logger.debug(f"Creating backtest export {name} to {s3_uri}")
        backtest_arn = self.aws_handler.forecast.create_predictor_backtest_export_job(
            PredictorBacktestExportJobName=name,
            PredictorArn=self.arn,
            Destination={
                "S3Config": {
                    "Path": s3_uri,
                    "RoleArn": self.aws_handler.ROLE_ARN,
                }
            },
            Tags=[
                {
                    "Key": "Name",
                    "Value": self.aws_handler.team,
                },
            ],
        )["PredictorBacktestExportJobArn"]

        logger.info("Waiting for predictor metrics to be ready...")
        try:
            wait_for_resource(
                "predictor_backtest_export_job", backtest_arn, self.aws_handler
            )
        except ResourceError as e:  # pragma: no cover
            logger.error(
                f"The metrics could not be exported because an error {e} ocurred"
            )
            return

        # Download the exported csv files and merge them into the destination path.
        logger.debug("Downloading and merging files")
        download_and_merge(
            s3_uri + "/accuracy-metrics-values/",
            os.path.join(destination_dir, "backtest_metrics.csv"),
            self.aws_handler,
        )
        download_and_merge(
            s3_uri + "/forecasted-values/",
            os.path.join(destination_dir, "backtest_forecasts.csv"),
            self.aws_handler,
        )

        return destination_dir

    def forecast(
        self,
        destination_path: str,
        quantiles: list = None,
    ):

        quantiles = quantiles or ["0.1", "0.5", "0.9"]

        name = "forecast_" + str(datetime.now().timestamp()).replace(".", "_")
        s3_uri = f"{self.aws_handler.s3_uri}/forecasts/{name}"

        # Wait for all datasets to be ready before forecasting.
        logger.info("Waiting for all datasets to be ready")
        for dataset_arn in self.aws_handler._datasets.values():
            try:
                wait_for_resource("dataset", dataset_arn, self.aws_handler)
            except ResourceError as e:  # pragma: no cover
                logger.error(f"Error in dataset {dataset_arn}: {e}")
                logger.error(
                    """The forecast could not be created because some of the datasets in the project contain errors.
                    You should re-upload the ones that failed."""
                )
                return

        # Wait for the predictor to be active before trying to create a forecast.
        logger.info("Waiting for the predictor to be ready")
        try:
            wait_for_resource("predictor", self.arn, self.aws_handler)
        except ResourceError as e:  # pragma: no cover
            logger.error(
                f"The forecast could not be generated because there was an error {e} when creating the predictor"
            )
            return

        logger.debug(f"Creating forecast {name}")
        forecast_arn = self.aws_handler.forecast.create_forecast(
            ForecastName=name,
            PredictorArn=self.arn,
            ForecastTypes=quantiles,
            Tags=[{"Key": "Name", "Value": self.aws_handler.team}],
        )["ForecastArn"]

        # Wait for forecast to be active before trying to make an export.
        logger.info("Waiting for forecast to be ready...")
        try:
            wait_for_resource("forecast", forecast_arn, self.aws_handler)
        except ResourceError as e:  # pragma: no cover
            logger.error(
                f"The forecast could not be created because there was an error {e} while creating the export"
            )
            return

        # Create the export
        logger.debug(f"Creating forecast export {name}__e to {s3_uri}")
        export_arn = self.aws_handler.forecast.create_forecast_export_job(
            ForecastExportJobName=name + "__e",
            ForecastArn=forecast_arn,
            Destination={
                "S3Config": {
                    "Path": s3_uri,
                    "RoleArn": self.aws_handler.ROLE_ARN,
                }
            },
            Tags=[{"Key": "Name", "Value": self.aws_handler.team}],
        )["ForecastExportJobArn"]

        ## Download exports and join them into the destination csv file.

        # Wait for forecast export to be active before trying download files.
        logger.debug(f"Waiting for forecast export {export_arn} to be ready...")
        try:
            wait_for_resource("forecast_export_job", export_arn, self.aws_handler)
        except ResourceError as e:  # pragma: no cover
            logger.error(
                f"The forecast could not be retrieved because there was an error {e} while creating the forecast"
            )
            return

        # Download the exported csv files and merge them into the destination path.
        logger.debug("Downloading and merging files")
        download_and_merge(s3_uri, destination_path, self.aws_handler)

        # Delete forecast since there is a quota.
        logger.debug(f"Deleting forecast with ARN {forecast_arn}")
        self.aws_handler.forecast.delete_forecast(ForecastArn=forecast_arn)
